src/proc/interface.rs

```diff
@@ -1,34 +1,41 @@
 use crate::arena::{Arena, Handle};
 
-struct Interface<'a> {
+pub struct Interface<'a, T> {
     expressions: &'a Arena<crate::Expression>,
-    uses: Vec<crate::GlobalUse>,
+    visitor: &'a mut T,
 }
 
-impl<'a> Interface<'a> {
-    fn add_inputs(&mut self, handle: Handle<crate::Expression>) {
+pub trait Visitor {
+    fn visitor_expr(&mut self, _: &crate::Expression) {}
+    fn visitor_lhs_expr(&mut self, _: &crate::Expression) {}
+}
+
+impl<'a, T> Interface<'a, T>
+where
+    T: Visitor,
+{
+    fn traverse_expr(&mut self, handle: Handle<crate::Expression>) {
         use crate::Expression as E;
+
+        self.visitor.visitor_expr(&self.expressions[handle]);
```

Comment:
can we assign this reference to a temporary to avoid indexing by `handle` twice?